sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala

```diff
@@ -279,4 +279,21 @@ object DataSourceUtils extends PredicateHelper {
       dataFilters.flatMap(extractPredicatesWithinOutputSet(_, partitionSet))
     (ExpressionSet(partitionFilters ++ extraPartitionFilter).toSeq, dataFilters)
   }
+
+  /**
+   * Determines whether a filter should be pushed down to the data source or not.
+   *
+   * @param expression The filter expression to be evaluated.
+   * @return A boolean indicating whether the filter should be pushed down or not.
+   */
+  def shouldPushFilter(expression: Expression): Boolean = {
+    expression.deterministic && !expression.exists {
+      case childExpression @ (_: Attribute | _: GetStructField) =>
```

Comment:
```suggestion
      case e @ (_: Attribute | _: GetStructField) =>
```
`.exists` visits all nodes in the tree, not only the children of the root node.